typing/typecore.ml

```diff
@@ -1941,17 +1941,30 @@ let check_univars env expans kind exp ty_expected vars =
   raise (Error (exp.exp_loc, env,
                 Less_general(kind, [Unification_trace.diff ty ty_expected])))
 
-(* Check that a type is not a function *)
-let check_application_result env statement exp =
-  let loc = exp.exp_loc in
-  match (expand_head env exp.exp_type).desc with
-  | Tarrow _ ->
-      Location.prerr_warning exp.exp_loc Warnings.Partial_application
-  | Tvar _ -> ()
-  | Tconstr (p, _, _) when Path.same p Predef.path_unit -> ()
-  | _ ->
-      if statement then
-        Location.prerr_warning loc Warnings.Statement_type
+let check_partial_application exp =
+  let rec f delay =
+    match (expand_head exp.exp_env exp.exp_type).desc, exp.exp_desc with
+    | Tarrow _, Texp_ident _ -> ()
```

Comment:
```suggestion
    | _, Texp_ident _ -> ()
```
(even in the Tvar case, we don't need to register a delayed check when the expression in a Texp_ident)

Reply:
I pushed some commits doing both of these things, together with some tests. Thanks!